flink-runtime/src/test/java/org/apache/flink/runtime/io/network/netty/NettyMessageClientDecoderDelegateTest.java

```diff
@@ -0,0 +1,339 @@
+
+		inputGate = createSingleInputGate(1);
+		RemoteInputChannel inputChannel = createRemoteInputChannel(
+			inputGate,
+			new TestingPartitionRequestClient(),
+			networkBufferPool);
+		inputGate.assignExclusiveSegments();
+		inputChannel.requestSubpartition(0);
+		handler.addInputChannel(inputChannel);
+		inputChannelId = inputChannel.getInputChannelId();
+
+		SingleInputGate releasedInputGate = createSingleInputGate(1);
+		RemoteInputChannel releasedInputChannel = new InputChannelBuilder()
+			.setMemorySegmentProvider(networkBufferPool)
+			.buildRemoteAndSetToGate(inputGate);
+		releasedInputGate.close();
+		handler.addInputChannel(releasedInputChannel);
+		releasedInputChannelId = releasedInputChannel.getInputChannelId();
+	}
+
+	@After
+	public void tearDown() throws IOException {
+		inputGate.close();
+
+		networkBufferPool.destroyAllBufferPools();
+		networkBufferPool.destroy();
+
+		channel.close();
+	}
+
+	/**
+	 * Verifies that the client side decoder works well for unreleased input channels.
+	 */
+	@Test
+	public void testDownstreamMessageDecode() throws Exception {
+		testNettyMessageClientDecoding(false, false, false);
+	}
+
+	/**
+	 * Verifies that the client side decoder works well for empty buffers. Empty buffers should not
+	 * consume data buffers of the input channels.
+	 */
+	@Test
+	public void testDownstreamMessageDecodeWithEmptyBuffers() throws Exception {
+		testNettyMessageClientDecoding(true, false, false);
+	}
+
+	/**
+	 * Verifies that NettyMessageDecoder works well with buffers sent to a released input channel.
+	 * The data buffer part should be discarded before reading the next message.
+	 */
+	@Test
+	public void testDownstreamMessageDecodeWithReleasedInputChannel() throws Exception {
+		testNettyMessageClientDecoding(false, true, false);
+	}
+
+	/**
+	 * Verifies that NettyMessageDecoder works well with buffers sent to a removed input channel.
+	 * The data buffer part should be discarded before reading the next message.
+	 */
+	@Test
+	public void testDownstreamMessageDecodeWithRemovedInputChannel() throws Exception {
+		testNettyMessageClientDecoding(false, false, true);
+	}
+
+	//------------------------------------------------------------------------------------------------------------------
+
+	private void testNettyMessageClientDecoding(
+		boolean hasEmptyBuffer,
+		boolean hasBufferForReleasedChannel,
+		boolean hasBufferForRemovedChannel) throws Exception {
+
+		ByteBuf[] encodedMessages = null;
+		List<NettyMessage> decodedMessages = null;
+		try {
+			List<NettyMessage> messages = createMessageList(
+				hasEmptyBuffer,
+				hasBufferForReleasedChannel,
+				hasBufferForRemovedChannel);
+
+			encodedMessages = encodeMessages(messages);
+
+			List<ByteBuf> partitionedBuffers = repartitionMessages(encodedMessages);
+			decodedMessages = decodeMessages(channel, partitionedBuffers);
+			verifyDecodedMessages(messages, decodedMessages);
+		} finally {
+			if (encodedMessages != null) {
```

Comment:
Check whether the passed buffers null inside `releaseBuffers` then we can remove this condition here.